### ISSUE TITLE
[java] LexException with unicode char in private static final String definition

### ISSUE THREAD

**Issue by rbri** [pmd/pmd#5093]
**Affects PMD Version:** 
PMD version: 7.3.0

**Description:**

switching from 6.55 to 7.3.0 shows several LexException errors.
The error output looks like this

    ...git_repo\htmlunit\src\main\java\org\htmlunit\html\DomText.java: LexException: Lexical error in file '...\git_repo\htmlunit\src\main\java\org\htmlunit\html\DomText.java' at line 197, column 23: "u" (117), after : "\'\\" (in lexical state DEFAULT)

And this is the code line 

    private static final String WHITESPACE_CHARS = " \t\r\n\u000C";

**Exception Stacktrace:**
none

**Code Sample demonstrating the issue:**
I have several issue like this with this code base - https://github.com/HtmlUnit/htmlunit

**Steps to reproduce:**
**Running PMD through:** *Maven* / *JDK 11*

**Comment by adangel** [comment]
Thanks for reporting. However, I can't seem to be able to reproduce this LexException.

> ...git_repo\htmlunit\src\main\java\org\htmlunit\html\DomText.java: LexException: Lexical error in file '...\git_repo\htmlunit\src\main\java\org\htmlunit\html\DomText.java' at line 197, column 23: "u" (117), after : "\'\\" (in lexical state DEFAULT)

[DomText.java](https://github.com/HtmlUnit/htmlunit/blob/master/src/main/java/org/htmlunit/html/DomText.java) seems to be parsed fine in the Designer.
And Line 197 looks like this:
https://github.com/HtmlUnit/htmlunit/blob/491dff87204bde0cdb63e5df723c72cee133757c/src/main/java/org/htmlunit/html/DomText.java#L197
> ```java
>     return (c < '\uE000' || c > '\uF8FF') && (c == ' ' || !Character.isWhitespace(c));
> ```

Your mentioned code line is in [DOMTokenList.java](https://github.com/HtmlUnit/htmlunit/blob/master/src/main/java/org/htmlunit/javascript/host/dom/DOMTokenList.java) on line 42: https://github.com/HtmlUnit/htmlunit/blob/491dff87204bde0cdb63e5df723c72cee133757c/src/main/java/org/htmlunit/javascript/host/dom/DOMTokenList.java#L42
> ```java
>     private static final String WHITESPACE_CHARS = " \t\r\n\u000C";
> ```

But this file also parses fine.


I've noticed, you have excluded these two files (https://github.com/HtmlUnit/htmlunit/blob/491dff87204bde0cdb63e5df723c72cee133757c/pmd-ruleset.xml#L14-L18). But even, if I comment out the excludes, I don't see any errors when I run `mvn pmd:check`.

Can you verify, whether this is still an issue?

**Comment by rbri** [comment]
@adangel thanks for working on this

![image](https://github.com/pmd/pmd/assets/2544132/c3c61142-abef-4dca-b5d2-e1cdc103b5f1)

As you can see, the problem is still there for me; i'm using jdk11 and simply did a mvn pmd:pmd.
For this test i have removed the two files from the exclude list.

**Comment by rbri** [comment]
pmd:check seems to produce the same output
![image](https://github.com/pmd/pmd/assets/2544132/2d33f191-8d19-4319-b764-c820cab34101)

**Comment by adangel** [comment]
@rbri I still had no luck reproducing the LexExceptions so far. The unicode escapes are supposed to be replaced already before we run the lexer (the lexer never sees the \u-escapes, only the final characters). For some reason, this doesn't seem to happen for you and the original source code is passed on to the lexer/parser.

Can you maybe run `mvn pmd:pmd -X > pmd-debug.txt` and post/attach the debug output? Maybe we see here some exceptions, errors - something that gives a hint what is going on.

The other thing that might be wrong is the classpath of maven-pmd-plugin - when overriding the PMD version, you need to override all pmd-related dependencies, see https://maven.apache.org/plugins/maven-pmd-plugin/examples/upgrading-PMD-at-runtime.html - if you also add overrides for "pmd-javascript" and "pmd-jsp", maybe this helps. Currently, pmd-core:7.3.0, pmd-java:7.3.0, pmd-javascript:7.0.0 and pmd-jsp:7.0.0 are on the classpath, so a mixture of 7.0.0 and 7.3.0. But this doesn't cause a problem for me.

Maybe it's cause by some stale left-over files. Does the problem go away after `mvn clean test-compile pmd:pmd`?

**Comment by rbri** [comment]
[pmd-debug.zip](https://github.com/user-attachments/files/16101748/pmd-debug.zip)
voila

**Comment by rbri** [comment]
> Does the problem go away after mvn clean test-compile pmd:pmd?

No

**Comment by rbri** [comment]
adding

                    <dependency>
                        <groupId>net.sourceforge.pmd</groupId>
                        <artifactId>pmd-javascript</artifactId>
                        <version>${pmd.version}</version>
                    </dependency>
                    <dependency>
                        <groupId>net.sourceforge.pmd</groupId>
                        <artifactId>pmd-jsp</artifactId>
                        <version>${pmd.version}</version>
                    </dependency>

to the pom does also not help

**Comment by rbri** [comment]
the problem goes away when switching from ojdkbuild\jdk-11.0.2 
to

openjdk version "17.0.2" 2022-01-18
OpenJDK Runtime Environment (build 17.0.2+8-86)
OpenJDK 64-Bit Server VM (build 17.0.2+8-86, mixed mode, sharing)

**Comment by rbri** [comment]
mvn clean compile pmd:pmd -DskipTests

works also with jdk8 (i only need 11 for compiling/running the tests)

**Comment by rbri** [comment]
[pmd_jdk8.zip](https://github.com/user-attachments/files/16101827/pmd_jdk8.zip)

this time with the 'enhanced' pom

**Comment by rbri** [comment]
[pmd-jdk17.zip](https://github.com/user-attachments/files/16101845/pmd-jdk17.zip)

**Comment by adangel** [comment]
So, I could kind of reproduce it a bit. The problem doesn't appear always. It has nothing to do with operating system or encoding or maven version (appears not only under Windows, but also under Linux). However, it appears only with older versions of Java 11.

More so, it is a heisenbug: it disappears as soon as you turn on the debugger. You can set a breakpoint at the exception in [JjtreeParserAdapter](https://github.com/pmd/pmd/blob/fc128c18ddbfa9dc4257abe0c36ba8576ce8f8d4/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/javacc/JjtreeParserAdapter.java#L38), but if you reset the frame and let the file parse again, it works.
After adding some stdout debugging info, I've seen the following:
- For the failing files (DomText.java and DOMTokenList.java), the [translated TextDocument](https://github.com/pmd/pmd/blob/fc128c18ddbfa9dc4257abe0c36ba8576ce8f8d4/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/javacc/CharStream.java#L41) is the same - means, there is no FragmentedTextDocument. The parsers works on the original TextDocument, which doesn't have the unicode sequences replaced.
- This happens only, if there were no fragment detected, when [the translated document is built](https://github.com/pmd/pmd/blob/fc128c18ddbfa9dc4257abe0c36ba8576ce8f8d4/pmd-core/src/main/java/net/sourceforge/pmd/lang/document/FragmentedDocBuilder.java#L62-L65).
- In [gobbleMaxWithoutEscape()](https://github.com/pmd/pmd/blob/fc128c18ddbfa9dc4257abe0c36ba8576ce8f8d4/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/javacc/BackslashEscapeTranslator.java#L33-L51), we try to read up until the next backslash. But for these two files, it seems, we read up until the end of the file and don't find any backslashes. Usually, this method is called multiple times, roughly once per fragment, util the whole file is read. But for the two problematic files, we _do not detect backslashes_.  We end up in [that if statement](https://github.com/pmd/pmd/blob/fc128c18ddbfa9dc4257abe0c36ba8576ce8f8d4/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/javacc/BackslashEscapeTranslator.java#L45-L48) with seenBackslash==false and we advance this.bufpos with the new offset, which is the end of the file.
- This method overrides [gobbleMaxWithoutEscape in EscapeTranslator](https://github.com/pmd/pmd/blob/fc128c18ddbfa9dc4257abe0c36ba8576ce8f8d4/pmd-core/src/main/java/net/sourceforge/pmd/lang/ast/impl/javacc/EscapeTranslator.java#L120-L123). However, the overridden method is indeed being called (I see the stdout print statements), **but it doesn't seem to be executed correctly**.

My suspicion is towards JIT compilation/optimization problems. And I kind of can see it: if I run the JVM with `-Xint` (interpreter mode only), it is slow, but correct. More precisely, if the optimization of the gobbleMaxWithoutEscape method is disabled (`export MAVEN_OPTS="-XX:CompileCommand=exclude,net.sourceforge.pmd.lang.ast.impl.javacc.BackslashEscapeTranslator::gobbleMaxWithoutEscape -XX:CompileCommand=exclude,net.sourceforge.pmd.lang.ast.impl.javacc.EscapeTranslator::gobbleMaxWithoutEscape -XX:CompileCommand=dontinline,net.sourceforge.pmd.lang.ast.impl.javacc.EscapeTranslator::gobbleMaxWithoutEscape -XX:CompileCommand=exclude,net.sourceforge.pmd.lang.ast.impl.javacc.EscapeTranslator::translateImpl -XX:CompileCommand=dontinline,net.sourceforge.pmd.lang.ast.impl.javacc.EscapeTranslator::translateImpl"`), then it also works correctly.

Since this problem is not reliably reproducible, I can only guess, that this is a JIT problem.

Here are some test results: I ran `mvn pmd:pmd` repeatedly until failure with `count=0; while [ $? -eq 0 ]; do count=$((count + 1)); echo "count: $count"; mvn --show-version --offline -Dpmd.skipPmdError=false pmd:pmd; done; echo "count: $count"`


**The broken JDK 11**
This is java 11.0.2+9, you can get this from https://jdk.java.net/archive/
With that version, it takes up to 5 runs, until it fails. Sometimes, only one processing error appears, sometimes two processing errors.
I needed 3 to 5 runs, until failure. So, it is not reliable: Only after 4 successful pmd runs, one run failed. :red_circle: 
```
openjdk 11.0.2 2019-01-15
OpenJDK Runtime Environment 18.9 (build 11.0.2+9)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.2+9, mixed mode)
```

I tried various other builds/versions:
* Oracle JDK 11 (from oracles archives): java 11.0.2 2019-01-15 LTS
Java(TM) SE Runtime Environment 18.9 (build 11.0.2+9-LTS)
Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.2+9-LTS, mixed mode): 1 of 5 runs failed. :red_circle: 
* AdoptOpenJDK 11: https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/tag/jdk-11.0.4%2B11
openjdk 11.0.4 2019-07-16
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.4+11)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.4+11, mixed mode): 1 of 3 runs failed. :red_circle: 
* AdoptOpenJDK 11: https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/tag/jdk-11.0.7%2B10
openjdk 11.0.7 2020-04-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.7+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.7+10, mixed mode): 1 of 1 runs failed. :red_circle: 
* AdoptOpenJDK 11: https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/tag/jdk-11.0.8%2B10
openjdk 11.0.8 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.8+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode): 1 of 6 runs failed. :red_circle: 
* AdoptOpenJDK 11: https://github.com/AdoptOpenJDK/openjdk11-binaries/releases/tag/jdk-11.0.9.1%2B1
openjdk 11.0.9.1 2020-11-04
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.9.1+1)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.9.1+1, mixed mode): 51 runs and 79 runs without failure :green_circle: 
* The latest Adoptium 11: openjdk 11.0.23 2024-04-16
OpenJDK Runtime Environment Temurin-11.0.23+9 (build 11.0.23+9)
OpenJDK 64-Bit Server VM Temurin-11.0.23+9 (build 11.0.23+9, mixed mode): 50 runs without failured :green_circle: 


In summary: It seems to have been fixed with 11.0.9.1+1 - that's the first version, that doesn't show this problem.
The problem also doesn't show with the latest Java 11.0.23+9

I tried to look at the differences https://github.com/openjdk/jdk11u/compare/jdk-11.0.8+10...jdk-11.0.9.1+1 but couldn't make out an obvious change, that would explain this.

**Comment by rbri** [comment]
Wow  will update my jdk11. Sorry for eating up so much of your time. 
Many thanks for your detailed analysis.

**Comment by rbri** [comment]
Reminds me of this... https://github.com/HtmlUnit/htmlunit-neko/pull/98

**Comment by rbri** [comment]
@adangel 
back form vacation - have updated to the latest JDK11 and it works without problem. Thanks a lot for your support.
